Sources/NIOPosix/SelectableEventLoop.swift

```diff
@@ -421,46 +427,80 @@ Further information:
         }
     }
 
+    private func run(_ task: UnderlyingTask) {
+        /* for macOS: in case any calls we make to Foundation put objects into an autoreleasepool */
+        withAutoReleasePool {
+            switch task {
+            case .function(let function):
+                function()
+            #if compiler(>=5.9)
+            case .unownedJob(let erasedUnownedJob):
+                if #available(macOS 14.0, iOS 17.0, watchOS 10.0, tvOS 17.0, *) {
+                    erasedUnownedJob.unownedJob.runSynchronously(on: self.asUnownedSerialExecutor())
+                } else {
+                    fatalError("Tried to run an UnownedJob without runtime support")
+                }
+            #endif
+            }
+        }
+    }
+
     /// Start processing I/O and tasks for this `SelectableEventLoop`. This method will continue running (and so block) until the `SelectableEventLoop` is closed.
     internal func run() throws {
         self.preconditionInEventLoop()
         defer {
-            var scheduledTasksCopy = ContiguousArray<ScheduledTask>()
             var iterations = 0
+            var drained = false
             repeat { // We may need to do multiple rounds of this because failing tasks may lead to more work.
-                scheduledTasksCopy.removeAll(keepingCapacity: true)
 
-                self._tasksLock.withLock { () -> Void in
+                let (scheduledTasksCopy, immediateTasksCopy) = self._tasksLock.withLock {
                     // In this state we never want the selector to be woken again, so we pretend we're permanently running.
                     self._pendingTaskPop = true
 
+                    // During the shutdown we don't need to be that careful about reallocs
+                    var scheduledTasksCopy = ContiguousArray<ScheduledTask>()
                     // reserve the correct capacity so we don't need to realloc later on.
                     scheduledTasksCopy.reserveCapacity(self._scheduledTasks.count)
                     while let sched = self._scheduledTasks.pop() {
                         scheduledTasksCopy.append(sched)
                     }
+                    var immediateTasksCopy = ContiguousArray<UnderlyingTask>()
+                    immediateTasksCopy.reserveCapacity(self._immediateTasks.count)
+                    while let immediate = self._immediateTasks.popFirst() {
+                        immediateTasksCopy.append(immediate)
+                    }
+                    return (scheduledTasksCopy, immediateTasksCopy)
                 }
 
+                // Run all the immediate tasks. They're all "expired" and don't have failFn,
+                // therefore the best course of action is to run them.
+                for task in immediateTasksCopy {
+                    self.run(task)
+                }
```

Comment:
This is a fairly substantial behavioural change. Previously these tasks got dropped. Normally I’d say the logic here is good but we should prepare ourselves for a few subtle bugs popping out of this.

Reply:
Yes, true. But without this, the application would have their `Task`-s/functions not executed at all. I don't think any behaviour can be worse than that. We haven't been noticing, because before `Tasks` this wasn't frequently exercised path. But all Tasks are submitted without `failFn`. But yes, there could be some unintended consequences.

Reply:
Agreed it's a substantial change but it's definitely a good one. I think this will actually make tearing down an EventLoop much easier than it was before this patch. Previously we just couldn't implement this sensibly, now we can.

But of course, it'll likely trigger a few subtle changes here or there but I can't see this changing anything that was reliably working fine for the worse.